plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,303 @@
+
+        hashValue = await doHash(runner.hub.db, timestampMs, timezone, teamId, ip, host, userAgent, numIdentifies2)
+
+        // set the distinct id to the new hash value
+        event.properties[`$anon_distinct_id`] = hashValue
+    } else if (event.distinct_id === SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID) {
+        hashValue = await doHash(runner.hub.db, timestampMs, timezone, teamId, ip, host, userAgent, numIdentifies)
+        // event before identify has been called, distinct id is the sentinel and needs to be replaced
+        event.distinct_id = hashValue
+        event.properties[`$distinct_id`] = hashValue
+    } else {
+        // this event is after identify has been called, so subtract 1 from the numIdentifies
+        hashValue = await doHash(runner.hub.db, timestampMs, timezone, teamId, ip, host, userAgent, numIdentifies - 1)
+    }
+
+    const sessionRedisKey = `cookieless_s:${hashValue}`
+    // do we have a session id for this user already?
+    let sessionInfo = await runner.hub.db.redisGet<{ s: string; t: number } | null>(
+        sessionRedisKey,
+        null,
+        'cookielessServerHashStep',
+        {
+            jsonSerialize: true,
+        }
+    )
+    // if not, or the TTL has expired, create a new one. Don't rely on redis TTL, as ingestion lag could approach the 30-minute session inactivity timeout
+    if (!sessionInfo || timestampMs - sessionInfo.t > 60 * 30 * 1000) {
+        const sessionId = new UUID7(timestampMs).toString()
+        sessionInfo = { s: sessionId, t: timestampMs }
+        await runner.hub.db.redisSet(sessionRedisKey, sessionInfo, 'cookielessServerHashStep', SESSION_TTL_SECONDS)
+    } else {
+        // otherwise, update the timestamp
+        await runner.hub.db.redisSet(
+            sessionRedisKey,
+            { s: sessionInfo.s, t: timestampMs },
+            'cookielessServerHashStep',
+            SESSION_TTL_SECONDS
+        )
+    }
+
+    event.properties['$session_id'] = sessionInfo.s
+
+    return [event]
+}
+
+function getProperties(
+    event: PluginEvent,
+    timestamp: string
+): {
+    userAgent: string | undefined
+    ip: string | undefined
+    host: string | undefined
+    timezone: string
+    timestampMs: number
+    teamId: number
+} {
+    const userAgent = event.properties?.['$raw_user_agent']
+    const ip = event.properties?.['$ip']
+    const host = event.properties?.['$host']
+    const timezone = event.properties?.['$timezone'] || TIMEZONE_FALLBACK
+    const timestampMs = DateTime.fromISO(timestamp).toMillis()
+    const teamId = event.team_id
+
+    return { userAgent, ip, host, timezone, timestampMs, teamId }
+}
+
+const localSaltMap: Record<string, Uint32Array> = {}
+const mutex = new ConcurrencyController(1)
+
+export async function getSaltForDay(db: DB, timestamp: number, timeZone: string): Promise<Uint32Array> {
+    // get the day based on the timezone
+    const parts = new Intl.DateTimeFormat('en', {
+        timeZone,
```

Comment:
We are simply trusting `event.properties[$timezone]` but we know that's not reliable. Besides including stuff like `Etc/Unknown`, customers can choose to override that for their own use.

And, to make matters worse, this fails catastrophically if we attempt to create it with an invalid timezone.

![image](https://github.com/user-attachments/assets/3b460854-c269-43c0-9a56-cc5b134f6c4e)

We need a `try/catch` block here

Reply:
Very fair, I'll fix this

Reply:
Done